### DISCUSSION TITLE
No module named 'pymodbus.client.sync'

### DISCUSSION THREAD

**Discussion by maree-yo** [pymodbus-dev/pymodbus#1182] [Help] [answered]
Hello,
I have just installed Pymodbus on my laptop (windows 11) and raspberry pi. I am using Python 3.9.12.
Both devices had an issue with the “pymodbus.client.sync” when I tried to run a simple code. I cannot proceed because the error “No module named 'pymodbus.client.sync'”
I am wondering if there is any solution to that problem.
Thank you.

Cheers,
Maree

**Comment by janiversen** [comment]
Yes the simple solution is to read the documentation, you need to adapt your code to version 3, “sync” no longer exist.

**Reply by maree-yo** [reply]
Thanks. I am starting to use the REPL.

**Reply by madeinaus** [reply]
Unfortunately, the documentation at https://pymodbus-n.readthedocs.io/en/latest/readme.html#summary still shows:

from pymodbus.client.sync import ModbusTcpClient

I did not find that "reading the documentation" helped me in this case.

**Comment by janiversen** [comment, marked as answer]
try reading the correct documentation 

https://pymodbus.readthedocs.io/en/latest/readme.html#summary

**Reply by ImogenWren** [reply]
broken link

**Reply by janiversen** [reply]
It is allowed to think

https://pymodbus.readthedocs.io/en/latest/

That link was 11 month old.

**Reply by pubdc** [reply]
I had a working bit of python code and it is now rendered out-of-order due to lack of backward compatibility. The examples and documentation don't suffice as they assume an expert level of understanding. There is no clear step-by-step 'flow', there is not enough broad online support to justify such a major change imho. There is no how-to on how to translate old code to new. Very disappointed to see my working code being broken and it will take days maybe weeks to get it back to work.

**Comment by janiversen** [comment]
That is simply not correct !!! We have added very easy to follow examples:
- simple_async_client.py
- simple_sync_client.py

If you read the documentation, you will see these clearly outlines as "to-go" examples, and as a guide on how to use the library...the rest are marked "advanced" and for a good reason.

Just download/copy the simple exampleskl, change the configuration to match your device and run....how much easier do you expect it to be ???

There are no step-by-step "flow" going from one release to another, just read "API_CHANGES.rst", that tells you exactly what changed, but the changes is something you have to do yourself.

What do you mean there are not enough online support, you send your message 1 hour ago !!!! do you forget that we are volunteers, in typical project you have to wait days to get a reply.

Why on earth did you update pymodbus to a new version, if you are not prepared to update your code ????? NO library offer backwards compatibility over major versions, that is exactly why we (like a lot of others) have
- minor releases where compatibility is ensured
- bigger releases where you need to make smaller adjustments (typically change a parameter or two)
- major releases (which is your case) where the fundamental structure have changed.

You might think there is not enough broad support, we the maintainers see it quite differently, do you think we make the changes just to upset you ???

In my humble opinion you should be happy and show a little respect, considering how many hours we use free of charge, to make a library that many users can use (and actually do use).

**Reply by MAKOMO** [reply]
You are absolutely right!
